src/main/java/com/pinterest/secor/util/ProtobufUtil.java

```diff
@@ -130,22 +133,40 @@ public boolean isConfigured() {
      * @throws RuntimeException
      *             when there's problem decoding protobuf message
      */
-    public Message decodeJsonMessage(String topic, byte[] payload){
+    private static final JsonFormat.Parser jsonParser = JsonFormat.parser().ignoringUnknownFields();
+    public Message decodeJsonMessage(String topic, byte[] payload) throws InvalidProtocolBufferException {
         try {
             Method builderGetter = allTopics ? messageClassForAll.getDeclaredMethod("newBuilder") : messageClassByTopic.get(topic).getDeclaredMethod("newBuilder");
             com.google.protobuf.GeneratedMessageV3.Builder builder = (com.google.protobuf.GeneratedMessageV3.Builder) builderGetter.invoke(null);
-            JsonFormat.parser().ignoringUnknownFields().merge(new String(payload, Charsets.UTF_8), builder);
+            jsonParser.merge(new InputStreamReader(new ByteArrayInputStream(payload)), builder);
             return builder.build();
+        } catch (InvalidProtocolBufferException e){
+            throw e;
         } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
-            throw new RuntimeException("Error parsing protobuf message", e);
-        } catch (InvalidProtocolBufferException e) {
+            throw new RuntimeException("Error parsing JSON message", e);
+        } catch (IOException e) {
+            throw new RuntimeException("Error creating read stream for JSON message", e);
+        }
+    }
+
+    public Message decodeProtobufMessage(String topic, byte[] payload){
+        Method parseMethod = allTopics ? messageParseMethodForAll : messageParseMethodByTopic.get(topic);
+        try {
+            return (Message) parseMethod.invoke(null, payload);
+        } catch (IllegalArgumentException e) {
```

Comment:
I would collapse these with `|` and have same error message, error cases are close enough

Reply:
I agree, but this is how it is in pinterest's code, so would like to keep changes minimal for now